grpc/src/main/java/com/linecorp/armeria/internal/grpc/HttpStreamReader.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.grpc;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.common.http.HttpData;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpObject;
+
+import io.grpc.Decompressor;
+import io.grpc.DecompressorRegistry;
+import io.grpc.Status;
+
+/**
+ * A {@link Subscriber} to read HTTP messages and pass to GRPC business logic.
+ */
+public class HttpStreamReader implements Subscriber<HttpObject> {
+
+    private static final Logger logger = LoggerFactory.getLogger(HttpStreamReader.class);
+
+    private final DecompressorRegistry decompressorRegistry;
+    private final StatusListener statusListener;
+
+    @VisibleForTesting
+    public final ArmeriaMessageDeframer deframer;
+
+    @Nullable
+    private Subscription subscription;
+
+    public HttpStreamReader(DecompressorRegistry decompressorRegistry,
+                     ArmeriaMessageDeframer deframer,
+                     StatusListener statusListener) {
+        this.decompressorRegistry = decompressorRegistry;
+        this.deframer = deframer;
+        this.statusListener = statusListener;
```

Comment:
Yep - I always have trouble deciding between whether internal helpers, like this one, deserve them (e.g., we don't add to private helpers). Went ahead and added to all constructor arguments.